### PR TITLE
#4003: added register_pre_operation_hook and register_post_operation_hook

### PR DESCRIPTION
**Pull request by arakhmati**

### CODE REVIEW

File: ttnn/ttnn/operations/relational.py [hunk 1]
```diff
@@ -192,7 +192,7 @@ def relational_function(
         return output_tensor
 
     relational_function.__name__ = f"ttnn.{name}"
-    relational_function.__doc__ = f"""{name}(input_tensor_a: ttnn.Tensor, input_tensor_b: ttnn.Tensor, *, memory_config: ttnn.MemoryConfig = ttnn.DRAM_MEMORY_CONFIG) -> ttnn.Tensor
+    relational_function.function.__doc__ = f"""{name}(input_tensor_a: ttnn.Tensor, input_tensor_b: ttnn.Tensor, *, memory_config: ttnn.MemoryConfig = ttnn.DRAM_MEMORY_CONFIG) -> ttnn.Tensor
```
Comment (eyonland) [inline ttnn/ttnn/operations/relational.py:195]: Why was the "function" reference needed on the docs... not seeing it.

File: docs/source/ttnn/usage.rst [hunk 2]
```diff
@@ -277,3 +276,40 @@ The following environment variable can be set in order to completely disable the
     ttnn.print_l1_buffers()
 
     ttnn.close_device(device)
+
+
+
+13. Register pre- and/or post-operation hooks
+---------------------------------------------
```
Comment (eyonland) [inline docs/source/ttnn/usage.rst:283]: I noticed in general we are leaving off the "Why" for the usage.  Should we start including that?

File: ttnn/ttnn/operations/activation.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ def activation_function(
         return output_tensor
 
     activation_function.__name__ = f"ttnn.{(name)}"
-    activation_function.__doc__ = f"""{(name)}(input_tensor: ttnn.Tensor, *, memory_config: ttnn.MemoryConfig = ttnn.DRAM_MEMORY_CONFIG) -> ttnn.Tensor
+    activation_function.function.__doc__ = f"""{(name)}(input_tensor: ttnn.Tensor, *, memory_config: ttnn.MemoryConfig = ttnn.DRAM_MEMORY_CONFIG) -> ttnn.Tensor
```
Comment (eyonland) [inline ttnn/ttnn/operations/activation.py:75]: Why is .function. needed ?

**Review by eyonland — APPROVED** [review]

File: docs/source/ttnn/usage.rst (hunk 2 above)
Comment (arakhmati) [reply docs/source/ttnn/usage.rst:283]: I think we already include the why in the examples where it makes sense to do so. Other examples are more open-ended on the other hand and I don't think we should restrict people's imagination with our reasons. The examples themselves already show why something could be used

**Review by cfjchu — APPROVED** [review]

**Pull request merged**